pkg/dataobj/internal/dataset/value.go

```diff
@@ -146,13 +147,61 @@ func (v Value) String() string {
 // ByteSlice returns v's value as a byte slice. If v is not a string,
 // ByteSlice returns a byte slice of the form "VALUE_TYPE_T", where T is the
 // underlying type of v.
-func (v Value) ByteArray() []byte {
+func (v *Value) ByteArray() []byte {
 	if ba, ok := v.any.(bytearray); ok {
 		return unsafe.Slice(ba, v.num)
 	}
 	panic(fmt.Sprintf("dataset.Value type is %s, not %s", v.Type(), datasetmd.VALUE_TYPE_BYTE_ARRAY))
 }
 
+// Buffer returns the memory pointed to by this Value. The returned buffer will have a capacity of at least sz bytes & may be allocated if the existing buffer is too small.
+func (v *Value) Buffer(sz int) []byte {
+	if v.cap == 0 {
+		dst := make([]byte, sz)
+		v.any = (bytearray)(unsafe.SliceData(dst))
+		v.cap = uint64(cap(dst))
+		return dst
+	}
+
+	var dst []byte
+	// Depending on which type this value was previously used for dictates how we reference the memory.
+	switch v.any.(type) {
+	case stringptr:
+		dst = unsafe.Slice(v.any.(stringptr), int(v.cap))
+	case bytearray:
+		dst = unsafe.Slice(v.any.(bytearray), int(v.cap))
+	default:
+		panic("unknown value type in Value's 'any' field")
+	}
+
+	// Grow the buffer attached to this Value if necessary.
+	if v.cap < uint64(sz) {
+		dst = slicegrow.GrowToCap(dst, sz)
+		v.any = (bytearray)(unsafe.SliceData(dst))
+		v.cap = uint64(cap(dst))
+	}
+	return dst
+}
+
+// SetByteArray updates the value to point to the provided byte slice.
```

Comment:
This also updates the type of Value to be a ByteArray; I think we should document that.